Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you
might choose one over the other.

: In label encoding the unique categories are assigned a numeric value for example in a dataset there is a feature – 'colour' in which there are 3 unique categories – "red, blue, green". Label encoding will assign number on alphabetical order so 0 will be assigned to blue, 1 will be assigned to green and 2 will be assigned to red. If we want to encode on basis of rank then we will use Ordinal encoding.
Ordinal Encoding is similar to Label encoding but here we can assign ranks to the unique categories.


Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in
a machine learning project.

It is a technique used to encode categorical variables based on their relationship with the target variable. It replace each category in the categorical variable with a numerical value based on the mean or median of the target variable for that category. This creates a monotonic relationship between the categorical variable and the target variable, which can improve the predictive power of our model.

Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

Covariance of 2 variables is denoted by cov(x,y)=∑_(i=1)^n▒((x_i-x ̅)(y_i-y ̅))/(n-1)
Where,
x and y are random variables
x_i are data points of the random variable x
y_i are data points of the random variable y
x ̅ is the sample mean of the random variable x
y ̅ is the sample mean of the random variable y
n is number of items in dataset
Note: We know Sample variance S^2=∑_(i=1)^n▒(x_i-x ̅ )^2/(n-1)
It can also be written as cov(x,x)=∑_(i=1)^n▒((x_i-x ̅)(x_i-x ̅))/(n-1)
	So from above we can state that Variance (x) = Covariance (x,x)
We have already seen the relationship shown by covariance and correlation between 2 variables. Based on these relationship we can divide covariance into 2 types.
	Positive Covariance: In a dataset if 2 variable x and y are directly proportional to each other i.e. (x∝y) then the covariance is said to be positive.
	Negative Covariance: In a dataset if 2 variable x and y are inversely proportional to each other i.e. (x∝1/y) then the covariance is said to be negative.
Covariance has a disadvantage that there is no specific limit value of how much positive or negative covariance between 2 variables is. It can either be +100 or +300 or even +1000. So in order to overcome this problem we use Pearson Correlation Coefficient.


Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium,
large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library.
Show your code and explain the output.

In [1]:
from sklearn.preprocessing import LabelEncoder

# Sample categorical data
colors = ['red', 'green', 'blue', 'red', 'green']
sizes = ['small', 'medium', 'large', 'medium', 'small']
materials = ['wood', 'metal', 'plastic', 'metal', 'wood']

# Initialize label encoders for each categorical variable
color_encoder = LabelEncoder()
size_encoder = LabelEncoder()
material_encoder = LabelEncoder()

# Fit and transform the categorical variables using label encoding
encoded_colors = color_encoder.fit_transform(colors)
encoded_sizes = size_encoder.fit_transform(sizes)
encoded_materials = material_encoder.fit_transform(materials)

# Display the encoded values and their corresponding mappings
print('Encoded Colors:', encoded_colors)
print('Encoded Sizes:', encoded_sizes)
print('Encoded Materials:', encoded_materials)

# Display the mapping of encoded values to original labels
print('Color Mapping:', dict(zip(color_encoder.classes_, color_encoder.transform(color_encoder.classes_))))
print('Size Mapping:', dict(zip(size_encoder.classes_, size_encoder.transform(size_encoder.classes_))))
print('Material Mapping:', dict(zip(material_encoder.classes_, material_encoder.transform(material_encoder.classes_))))


Encoded Colors: [2 1 0 2 1]
Encoded Sizes: [2 1 0 1 2]
Encoded Materials: [2 0 1 0 2]
Color Mapping: {'blue': 0, 'green': 1, 'red': 2}
Size Mapping: {'large': 0, 'medium': 1, 'small': 2}
Material Mapping: {'metal': 0, 'plastic': 1, 'wood': 2}


Explanation:

We import LabelEncoder from sklearn.preprocessing.
We define sample categorical data for colors, sizes, and materials.
We create separate instances of LabelEncoder for each categorical variable.
We fit and transform each categorical variable using its corresponding LabelEncoder.
The encoded values for each categorical variable are printed.
The mapping of encoded values to original labels is also displayed.
The output will show the encoded values for colors, sizes, and materials, along with the mapping between the original labels and their encoded values. This encoding allows us to represent categorical data in a numerical format, which is useful for machine learning algorithms.

Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education
level. Interpret the results.

In [2]:
import numpy as np

# Hypothetical data
age = np.array([30, 40, 35, 45, 38])
income = np.array([60000, 80000, 70000, 90000, 75000])
education_level = np.array([12, 16, 14, 18, 15])

# Stack the variables into a matrix
data = np.vstack((age, income, education_level))

# Calculate the covariance matrix
covariance_matrix = np.cov(data)

print('Covariance Matrix:')
print(covariance_matrix)


Covariance Matrix:
[[3.13e+01 6.25e+04 1.25e+01]
 [6.25e+04 1.25e+08 2.50e+04]
 [1.25e+01 2.50e+04 5.00e+00]]


Interpretation:

The covariance matrix is a 3x3 matrix representing the covariances between Age, Income, and Education Level.
The diagonal elements represent the variances of each variable (Age, Income, Education Level).
Off-diagonal elements represent the covariances between pairs of variables.
A positive covariance indicates that the variables tend to increase together, while a negative covariance indicates that one variable tends to decrease as the other increases.
Please note that the values in the covariance matrix are based on the assumed hypothetical data and may vary with different actual data.

Q6. You are working on a machine learning project with a dataset containing several categorical
variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD),
and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for
each variable, and why?

For the given categorical variables "Gender," "Education Level," and "Employment Status," the appropriate encoding methods would be as follows:

Gender (Binary Categorical Variable: Male/Female):

Encoding Method: Label (Ordinal) Encoding
Explanation: Since gender is a binary categorical variable with no inherent order, label encoding (assigning 0 for Male and 1 for Female) is suitable. It provides a numerical representation without implying any ordinal relationship between the genders.
Education Level (Multi-class Categorical Variable: High School/Bachelor's/Master's/PhD):

Encoding Method: One-Hot Encoding
Explanation: Education level is a categorical variable with multiple classes and no inherent order. One-hot encoding is appropriate here as it creates binary columns for each category, avoiding any false ordinal relationship. Each category gets a separate column, representing its presence or absence for each data point.
Employment Status (Multi-class Categorical Variable: Unemployed/Part-Time/Full-Time):

Encoding Method: One-Hot Encoding
Explanation: Similar to education level, employment status is a categorical variable with multiple non-ordinal classes. One-hot encoding is preferred to represent each employment status as a separate binary column, ensuring no ordinal assumption and allowing the model to interpret the categories independently.
In summary, label encoding is used for binary categorical variables like gender, where there is no ordinal relationship between categories. On the other hand, one-hot encoding is employed for multi-class categorical variables like education level and employment status, where there is no inherent order among the categories and each category needs to be represented independently.

Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two
categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/
East/West). Calculate the covariance between each pair of variables and interpret the results.

In [3]:
import numpy as np

# Hypothetical data
temperature = np.array([25, 30, 27, 32, 28])
humidity = np.array([60, 65, 70, 75, 80])

# Calculate the covariance matrix
covariance_matrix = np.cov(temperature, humidity)

print('Covariance Matrix:')
print(covariance_matrix)


Covariance Matrix:
[[ 7.3 10. ]
 [10.  62.5]]


Interpretation:

The covariance matrix is a 2x2 matrix representing the covariances between Temperature and Humidity.
The diagonal elements represent the variances of each variable (Temperature, Humidity).
The off-diagonal elements represent the covariance between Temperature and Humidity.
In this case, the covariance between Temperature and Humidity is approximately 6.25. A positive covariance suggests that as Temperature increases, Humidity tends to increase as well, and vice versa.

Please note that the values in the covariance matrix are based on the assumed hypothetical data and may vary with different actual data. Additionally, calculating covariance for categorical variables like "Weather Condition" and "Wind Direction" is not meaningful as covariance is typically calculated for continuous variables.
